### ISSUE TITLE
Unable to reset password if customer has address from not allowed country

### ISSUE THREAD

**Issue by ihor-sviziev** [magento/magento2#18170]
### Preconditions
1. Magento 2.2.5

### Steps to reproduce
1. Create customer on frontend, add some customer address with Germany (just an example) and log out
2. Go to Admin >> Stores >> Configuration >> General >> General
3. In Country Options >> Allow Countries deselect Germany (mark as not allowed), click to Save Config (probably flushing cache also required)
4. On frontend got to "Forgot password" form and try to reset password for customer that you created in step 1

### Expected result
1. Customer should get reset password link without any issues

### Actual result
1. Customer get's the error "We're unable to send the password reset email."
![image](https://user-images.githubusercontent.com/1873745/45864468-c139f800-bd82-11e8-8dc9-c31e2cc49f71.png)
2. In logs we have 
```
[2018-09-21 06:09:22] main.CRITICAL: Exception message: Invalid value of "GB" provided for the countryId field.
Trace: #0 /path_to_magento/magento2/vendor/magento/module-customer/Model/ResourceModel/CustomerRepository.php(266): Magento\Customer\Model\ResourceModel\AddressRepository->save(Object(Magento\Customer\Model\Data\Address))
#1 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Customer\Model\ResourceModel\CustomerRepository->save(Object(Magento\Customer\Model\Data\Customer), NULL)
#2 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->___callParent('save', Array)
#3 /path_to_magento/magento2/vendor/magento/module-customer/Model/Plugin/CustomerRepository/TransactionWrapper.php(44): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Customer\Model\Data\Customer), NULL)
#4 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(135): Magento\Customer\Model\Plugin\CustomerRepository\TransactionWrapper->aroundSave(Object(Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor), Object(Closure), Object(Magento\Customer\Model\Data\Customer))
#5 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Customer\Model\Data\Customer))
#6 /path_to_magento/magento2/generated/code/Magento/Customer/Model/ResourceModel/CustomerRepository/Interceptor.php(26): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->___callPlugins('save', Array, Array)
#7 /path_to_magento/magento2/vendor/magento/module-customer/Model/AccountManagement.php(1284): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->save(Object(Magento\Customer\Model\Data\Customer))
#8 /path_to_magento/magento2/generated/code/Magento/Customer/Model/AccountManagement/Interceptor.php(271): Magento\Customer\Model\AccountManagement->changeResetPasswordLinkToken(Object(Magento\Customer\Model\Data\Customer), 'e53fc7d535673c8...')
#9 /path_to_magento/magento2/vendor/magento/module-customer/Model/AccountManagement.php(567): Magento\Customer\Model\AccountManagement\Interceptor->changeResetPasswordLinkToken(Object(Magento\Customer\Model\Data\Customer), 'e53fc7d535673c8...')
#10 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Customer\Model\AccountManagement->initiatePasswordReset('ihor@example.com', 'email_reset', '4')
#11 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Customer\Model\AccountManagement\Interceptor->___callParent('initiatePasswor...', Array)
#12 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Customer\Model\AccountManagement\Interceptor->Magento\Framework\Interception\{closure}('ihor@example.com', 'email_reset')
#13 /path_to_magento/magento2/generated/code/Magento/Customer/Model/AccountManagement/Interceptor.php(91): Magento\Customer\Model\AccountManagement\Interceptor->___callPlugins('initiatePasswor...', Array, Array)
#14 /path_to_magento/magento2/vendor/magento/module-customer/Controller/Account/ForgotPasswordPost.php(76): Magento\Customer\Model\AccountManagement\Interceptor->initiatePasswordReset('ihor@example.com', 'email_reset')
#15 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Customer\Controller\Account\ForgotPasswordPost->execute()
#16 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->___callParent('execute', Array)
#17 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->Magento\Framework\Interception\{closure}()
#18 /path_to_magento/magento2/generated/code/Magento/Customer/Controller/Account/ForgotPasswordPost/Interceptor.php(26): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->___callPlugins('execute', Array, Array)
#19 /path_to_magento/magento2/vendor/magento/framework/App/Action/Action.php(107): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->execute()
#20 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\Action\Action->dispatch(Object(Magento\Framework\App\Request\Http))
#21 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->___callParent('dispatch', Array)
#22 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#23 /path_to_magento/magento2/generated/code/Magento/Customer/Controller/Account/ForgotPasswordPost/Interceptor.php(39): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->___callPlugins('dispatch', Array, Array)
#24 /path_to_magento/magento2/vendor/magento/framework/App/FrontController.php(55): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#25 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\FrontController->dispatch(Object(Magento\Framework\App\Request\Http))
#26 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\FrontController\Interceptor->___callParent('dispatch', Array)
#27 /path_to_magento/magento2/vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php(94): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#28 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(135): Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#29 /path_to_magento/magento2/vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php(73): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#30 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(135): Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#31 /path_to_magento/magento2/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#32 /path_to_magento/magento2/generated/code/Magento/Framework/App/FrontController/Interceptor.php(26): Magento\Framework\App\FrontController\Interceptor->___callPlugins('dispatch', Array, NULL)
#33 /path_to_magento/magento2/vendor/magento/framework/App/Http.php(135): Magento\Framework\App\FrontController\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#34 /path_to_magento/magento2/generated/code/Magento/Framework/App/Http/Interceptor.php(24): Magento\Framework\App\Http->launch()
#35 /path_to_magento/magento2/vendor/magento/framework/App/Bootstrap.php(256): Magento\Framework\App\Http\Interceptor->launch()
#36 /path_to_magento/magento2/pub/index.php(37): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
#37 {main} [] []
```

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@ihor-sviziev do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `G1 Passed` will be added to the issue automatically. Please, edit issue description if needed, until label `G1 Passed` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `G2 Passed` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `acknowledged` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report is acknowledged.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-95165`, `MAGETWO-95166` were created

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev. Thank you for your report. 
The issue has been fixed in magento/magento2#18179 by @dmytro-ch in 2.3-develop branch
Related commit(s):
 - [030b223efa3c7e5811ccddbf064d33fb7224c888](https://github.com/magento/magento2/commit/030b223efa3c7e5811ccddbf064d33fb7224c888)
 - [54b6382b4b8c142e99afa62382c6adc729e82adb](https://github.com/magento/magento2/commit/54b6382b4b8c142e99afa62382c6adc729e82adb)
 - [406d00003ee1190714ee90b53577a4eaf4c062ad](https://github.com/magento/magento2/commit/406d00003ee1190714ee90b53577a4eaf4c062ad)
 - [ebd39d36bb3422eec41a939f5c20a9b1ae2efde3](https://github.com/magento/magento2/commit/ebd39d36bb3422eec41a939f5c20a9b1ae2efde3)
 - [91275d606be59a30b6dc327309c7d7faa2cdefa5](https://github.com/magento/magento2/commit/91275d606be59a30b6dc327309c7d7faa2cdefa5)
 - [33e42acb3b8ebd42eec2e2752bbd33fd13a568db](https://github.com/magento/magento2/commit/33e42acb3b8ebd42eec2e2752bbd33fd13a568db)
 - [97632b368698137434a055bf6f8145fa9c355c65](https://github.com/magento/magento2/commit/97632b368698137434a055bf6f8145fa9c355c65)
 - [338d568cfdb7979cb644d9b561846bb281bc81a8](https://github.com/magento/magento2/commit/338d568cfdb7979cb644d9b561846bb281bc81a8)

The fix will be available with the upcoming 2.3.1 release.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev. Thank you for your report. 
The issue has been fixed in magento/magento2#19964 by @dmytro-ch in 2.2-develop branch
Related commit(s):
 - [317c02a344dac510619013a68eca167a5fd384b2](https://github.com/magento/magento2/commit/317c02a344dac510619013a68eca167a5fd384b2)

The fix will be available with the upcoming 2.2.8 release.
